patches/tModLoader/Terraria/ModLoader/ChestLoot.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Terraria.Enums;
+using Terraria.GameContent.Events;
+using Terraria.ID;
+using Terraria.Localization;
+
+namespace Terraria.ModLoader;
+
+public partial class ChestLoot {
+	private static readonly Item defaultInstance = default;
+
+	private readonly List<Entry> items;
+	private string name;
+
+	private readonly List<(int nextTo, bool after)> putCandidates;
+	private readonly List<Entry> putCandidates2; // list that contains all entries those going to get from putCandidates
+
+	public IReadOnlyList<Entry> Items {
+		get {
+			return items;
+		}
+	}
+
+	public Entry this[int item] {
+		get {
+			int index = items.FindIndex(x => x.item.type.Equals(item));
+			bool hasInNormal = index != -1;
+			if (hasInNormal)
+				return items[index];
+
+			index = putCandidates2.FindIndex(x => x.item.type.Equals(item));
+			return putCandidates2[index];
+		}
+	}
+
+	public Entry this[Index index] {
+		get {
+			var ind2 = items.ElementAtOrDefault(index);
+			bool hasInNormal = ind2.item != defaultInstance;
+			if (hasInNormal)
+				return ind2;
+
+			return putCandidates2.ElementAtOrDefault(index);
+		}
+	}
+
+	public ChestLoot() {
+		items = new();
+		putCandidates = new();
+		putCandidates2 = new();
+	}
+
+	public void RegisterShop(int npcId, string name = "Shop") {
+		this.name = name;
+		Main.TMLLootDB.RegisterNpcShop(npcId, this, name);
+	}
+
+	private void AddCandidates(List<Entry> entries) {
+		List<(int nextTo, bool after)> candidates = putCandidates;
+		List<Entry> candidates2 = putCandidates2;
+		candidates.Reverse();
+		candidates2.Reverse();
```

Comment:
This mutates the original lists... so calling `AddCandidates` multiple times will mess up